### PR TITLE
[SMD-120] Remove Unneeded FSHF Data from future ingests

### PR DESCRIPTION
**Pull request by Connor-OS**
Deal with the conditional formating in the funding profiles tab of the TF spreedsheet. There are a number of cells on this tab that are covered by a grey box dependant on a number of conditions on page. This PR is to drop these cells from the data extract going forward with round 3 and 4 data extracts.

There are 3 main considerations:
1. all cells for before 2020 are greyed out except for "Other funding sources"
2. if fund type is Town Deal and Programme only is selected the question: "How much of your CDEL forcast is contractually commited" is greyed out
3. if fund type is Future High Street Fund:
        a. Section B is not required
        b. "5% CDEL pre-payment", "RDEL payment", and "How much of your RDEL is commited" are not required
        c. Payment for reporting period is only required forcasted half a year ahead

See images bellow

All of this logic has been added at the end of extract_funding_questions() and extract_funding_data() to drop the relevant cells.

### Change description
_A brief description of the pull request_

- [ ] Unit tests and other appropriate tests added or updated (TODO)
- [ ] README and other documentation has been updated / added (if needed)
- [ ] Commit messages are meaningful and follow good commit message guidelines (e.g. "FS-XXXX: Add margin to nav items preventing overlapping of logo")

### How to test
_If manual testing is needed, give suggested testing steps_

### Screenshots of UI changes (if applicable)
basic

1.
before_2020

2.
programme_only

3. a)
section_b

3. b,c)
FHSF

### CODE REVIEW

File: core/extraction/towns_fund.py
Comment (Joel-Pearce) [inline core/extraction/towns_fund.py:441]: If a user is uploading for a FHSF we don't even need to run the function that extracts and transforms funding question data. We can skip this entire line of code:

```
    towns_fund_extracted["Funding Questions"] = extract_funding_questions(
        df_ingest["4a - Funding Profiles"],
        programme_id,
    )
```

File: core/extraction/towns_fund.py
Comment (Joel-Pearce) [inline core/extraction/towns_fund.py:571]: A user only uploads for either TD or FHSF. In the case where they have selected their fund_type as FHSF, cells pertaining to this financial information are greyed out. Rather than dropping this information after collecting it, it might make more sense to not collect it in the first place?

**Review by Joel-Pearce — CHANGES_REQUESTED** [review]
Couple of comments about the need to drop data when we can just not collect it in the first place, let me know if you have any questions

File: core/extraction/towns_fund.py
Comment (Joel-Pearce) [reply core/extraction/towns_fund.py:571]: In this case, that might be more difficult than what you've done, or more expensive; if that is the case then dropping might be easier :)

File: core/validation_schema.py
Comment (cyrusdobbs) [inline core/validation_schema.py:847]: This can be added for the Round 3 schema as well

File: core/extraction/towns_fund.py
Comment (Connor-OS) [reply core/extraction/towns_fund.py:441]: Yeah, I agree with this. Only issue is, correct me if I'm wrong, the validator expects an empty sheet with the correct columns to be there so we would need to return a dummy sheet and this was the simplest way I could implement that behaviour

File: core/extraction/towns_fund.py
Comment (Connor-OS) [reply core/extraction/towns_fund.py]: I did think about this. I think might be quite a bit harder, since we use iloc to grab the information of the page in the first instance, but we need to drop data based on some very specific rules, that would be hard to apply there.

File: core/extraction/towns_fund.py
Comment (nsnape) [inline core/extraction/towns_fund.py:575]: Looking at the form, this should only happen for the "mandatory" funding sources, not the custom ones. Need to adda bit of logic here, ie just remove these dates if the "funding source name" is one of the remaining 2 valid mandatory funding source names. Does that make sense?

File: core/extraction/towns_fund.py
Comment (nsnape) [inline core/extraction/towns_fund.py:579]: this line sort of gets done twice, might be cleaner to just do it once at the start of the function and set to a variable, ie `fund_type = project_lookup.values())[0].split("-")[0]`

File: core/extraction/towns_fund.py
Comment (nsnape) [inline core/extraction/towns_fund.py:582]: The only problem with this approach is that users could theoretically legitimately add, for example, "Town Deals 5% CDEL Pre-Payment" as a "custom" funding source name, and in this case should be able to enter these fields without them being dropped? I guess we could just not worry about that very unlikely edge-case? Might be worth checking with a BA?

File: core/extraction/towns_fund.py
Comment (nsnape) [reply core/extraction/towns_fund.py:441]: you could potentially add an optional param to the funding question extraction function, defaulted to False, something like `return_empty_sheet=False`, then add a line right at the start of the function that just returns an empty df with the correct columns? 

This would save on unnecessary computation.

File: tests/extraction_tests/resources/mock_sheet_data/start_page_mock.csv
Comment (nsnape) [inline tests/extraction_tests/resources/mock_sheet_data/start_page_mock.csv:6]: does this mean the tests have effectively been changed from round 3 to round 4?

Comment (nsnape) [reply tests/extraction_tests/resources/mock_sheet_data/start_page_mock.csv:6]: We should probably leave the data as is (testing round 3) and extend as necessary, on import, for further subsequent rounds. That way keep all the historical testing intact.

File: tests/extraction_tests/resources/mock_sheet_data/funding_profiles_mock.csv
Comment (nsnape) [inline tests/extraction_tests/resources/mock_sheet_data/funding_profiles_mock.csv:20]: I think if we go with this approach of have different mock data sources for HS and FHSF, we should at least split them into different sub-folders to keep it clearer.

Comment (nsnape) [reply tests/extraction_tests/resources/mock_sheet_data/funding_profiles_mock.csv:20]: same for assertion data

**Review by nsnape — CHANGES_REQUESTED** [review]

File: core/extraction/towns_fund.py
Comment (Connor-OS) [reply core/extraction/towns_fund.py:559]: The Yes/No is a drop down. So ideally users will get a dropdown error if they don't enter that field. I get that it is a bit of a hack/loose. Maybe there is a better way though

Comment (Connor-OS) [reply core/extraction/towns_fund.py:559]: Good spot about 25/26. I would have missed that

File: core/extraction/towns_fund.py
Comment (nsnape) [reply core/extraction/towns_fund.py:441]: Have you treid the Pndas .isempty method?
`assert df.empty`

Comment (nsnape) [reply core/extraction/towns_fund.py:441]: Or (apparently) even faster:
`assert len(df.index) > 0:`

File: tests/extraction_tests/resources/mock_sheet_data/start_page_mock.csv
Comment (Connor-OS) [reply tests/extraction_tests/resources/mock_sheet_data/start_page_mock.csv:6]: reverted, this in my latest push 👍

File: tests/extraction_tests/resources/mock_sheet_data/funding_profiles_mock.csv
Comment (Connor-OS) [reply tests/extraction_tests/resources/mock_sheet_data/funding_profiles_mock.csv:20]: I have now changed mock data to a single source of truth

File: core/extraction/towns_fund.py
Comment (nsnape) [inline core/extraction/towns_fund.py:500]: Although probably not a problem with a dict this small, a more efficient (better practice) way to do this is
`next(iter(project_lookup.values())).split("-")[0]`
as it doesn't then do the relatively expensive transaction of casting to a full list of which you don't actually need any value other than the first

File: tests/extraction_tests/test_round_three_extraction.py
Comment (nsnape) [inline tests/extraction_tests/test_round_three_extraction.py:340]: typo in forecast

File: tests/extraction_tests/test_round_three_extraction.py
Comment (nsnape) [inline tests/extraction_tests/test_round_three_extraction.py:359]: maybe could add a quick assertion to check that some rows are indeed extracted too, ie if start date is less than datetime(2023, 10, 1). Good to check the expected data as well as what has been removed

File: core/validation_schema.py
Comment (nsnape) [inline core/validation_schema.py:374]: I think you may have added this to Round 2 instead of Round 3

Comment (nsnape) [reply core/validation_schema.py:374]: Which also makes me think, whilst you're here could you re-order the schema's to appear in numerical order in the file? Makes more sense than the random order they're in now

File: core/extraction/towns_fund.py
Comment (nsnape) [reply core/extraction/towns_fund.py]: did we just decide that we'd leave this as is?

**Review by nsnape — CHANGES_REQUESTED** [review]

File: core/extraction/towns_fund.py
Comment (Connor-OS) [reply core/extraction/towns_fund.py]: I've added some conditional logic to ensure: df_funding["Funding Source Type"] == "Towns Fund" which should ensure that "Other funding sources" are unaffected

File: core/extraction/towns_fund.py
Comment (Connor-OS) [reply core/extraction/towns_fund.py:500]: Good to know, and added. I can see how this eis more effecient

File: core/validation_schema.py
Comment (nsnape) [reply core/validation_schema.py:409]: haha that's a hard change to check, as the schemas are quite similar, git has only detected partial changes :D It's too clever sometimes...

File: tests/extraction_tests/test_round_three_extraction.py
Comment (Connor-OS) [reply tests/extraction_tests/test_round_three_extraction.py:361]: Changed line 336 and 345 to assert on the expected length of each extracted dataframe 200, and 104 respectively

**Review by nsnape — COMMENTED** [review]
Ok, looking good. Let's test in in TEST env before we merge in though, I can approve when that's done and we have a strategy to update the data in place. Ta

**Review by nsnape — APPROVED** [review]

**Pull request merged**
